### PR TITLE
Sync stack - Netty file and stream request.

### PR DESCRIPTION
**Pull request by kasobol-msft**
In this PR:
- optimizations around file upload.
- attempt to handle stream without reactor.
- revisit String logic (to make sure encoding is correct)
- use a bit less allocating approach to upload bytes (this is still under investigation)

### CODE REVIEW

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java [hunk 1]
```diff
@@ -165,16 +176,59 @@
...
+                    (c, stream) -> {
+                        if (c.channel().pipeline().get(ChunkedWriteHandler.class) == null) {
+                            c.addHandlerLast("reactor.left.chunkedWriter", new ChunkedWriteHandler());
+                        }
+
+                        return new ChunkedStream(stream);
+                    },
+                    (stream) -> {
+                        try {
+                            stream.close();
```
Comment (alzimmermsft) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java:227]: Does this close the underlying `InputStream` from `InputStreamContent`? Generally, we've called out that our APIs taking `InputStream` do not close them.

File: sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java [hunk 2]
```diff
@@ -178,15 +178,13 @@ private static RequestBody toOkHttpRequestBodySynchronously(BinaryData bodyConte
         }
 
         BinaryDataContent content = BinaryDataHelper.getContent(bodyContent);
-        if (content instanceof ByteArrayContent) {
+        if (content instanceof ByteArrayContent || content instanceof StringContent) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java:181]: I'm guessing the old handler for `StringContent` did the same operations as `StringContent.toBytes()`?

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java (hunk 1 above)
Comment (kasobol-msft) [reply sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java:227]: yeah. it does. I'll change this.

File: sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java (hunk 2 above)
Comment (kasobol-msft) [reply sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java:181]: yeah, that's good point. I approximated what I found for netty here... I'll revert this.

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java [hunk 3]
```diff
@@ -176,6 +178,68 @@
...
+                        // Completed 7,720 operations in a weighted-average of 60.01s (128.65 ops/s, 0.008 s/op)
+                        return reactorNettyOutbound.sendByteArray(
+                            Mono.defer(() -> Mono.just(binaryDataContent.toBytes())));
+                    case 9:
+                        // Completed 7,391 operations in a weighted-average of 60.01s (123.16 ops/s, 0.008 s/op)
+                        // Completed 7,493 operations in a weighted-average of 60.00s (124.88 ops/s, 0.008 s/op)
+                        // Completed 7,589 operations in a weighted-average of 60.00s (126.47 ops/s, 0.008 s/op)
+                        // Completed 7,901 operations in a weighted-average of 60.01s (131.66 ops/s, 0.008 s/op)
+                        return reactorNettyOutbound.sendByteArray(
+                            Mono.fromCallable(binaryDataContent::toBytes));
```
Comment (kasobol-msft) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java]: I've run benchmark for this

# 1 MB Payload (~70 samples)

pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=1 (72)
avg 157.29&pm;3.85
max 163.84
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=2 (72)
avg 157.83&pm;3.89
max 164.64
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=4 (72)
avg 157.07&pm;3.62
max 164.29
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=5 (71)
avg 157.41&pm;3.63
max 163.91
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=6 (71)
avg 158.04&pm;3.74
max 163.90
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=7 (71)
avg 156.61&pm;3.62
max 163.64
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=8 (71)
avg 157.32&pm;3.48
max 163.80
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1048576 selection=9 (71)
avg 157.28&pm;3.90
max 163.82

# 1KB payload (100 sampleS)

pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=1 (100)
avg 13921.12&pm;156.85
max 14249.53
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=2 (100)
avg 13964.76&pm;145.93
max 14277.69
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=4 (100)
avg 13910.00&pm;118.84
max 14172.90
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=5 (100)
avg 13943.71&pm;131.69
max 14247.80
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=6 (100)
avg 13917.11&pm;148.29
max 14307.11
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=7 (100)
avg 13918.25&pm;156.27
max 14253.78
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=8 (100)
avg 13900.49&pm;136.12
max 14221.69
pipelinesend bytes netty 1.0.0-beta.1-jar-with-dependencies parallel=2 size=1024 selection=9 (100)
avg 13930.74&pm;143.81
max 14255.51

**Pull request merged**